### ISSUE TITLE
Promise.coroutine documentation

### ISSUE THREAD

**Issue by liushigit** [petkaantonov/bluebird#1426]
I was confused by [the doc](http://bluebirdjs.com/docs/api/promise.coroutine.html) for the `coroutine` method, which says the method

> Returns a function that can use yield to yield promises. 

My understanding of this wording is that, this method "returns a function that may (or may not) yield promises".

But in the provided code example, the returned function is only used as a plain function. There's no "yielded promises" from that function. If this is the example for the "may not" case, shall there be an example for the other case?

**Comment by benjamingr** [comment]
Yeah that documentation is silly. It _takes_ a function that can use yield to yield promises and _returns_ a promise returning function.

A PR would be welcome

**Comment by liushigit** [comment]
Here's the `coroutine` method's [source code](https://github.com/petkaantonov/bluebird/blob/master/src/generators.js#L191). 

The first sentence may be changed to "Takes a [`GeneratorFunction`](https://developer.mozilla.org/en/docs/Web/JavaScript/Guide/Iterators_and_Generators) producing a generator that yields promise(s), and returns a `Promise` returning function."

The returned `Promise` is defined at [this line(https://github.com/petkaantonov/bluebird/blob/master/src/generators.js#L43), but I don't understand its use.

**Comment by benjamingr** [comment]
@liushigit I'll clarify, I agree that the docs are confusing (and wrong), and if you want to submit a pull request I promise to review it and merge it.

**Comment by liushigit** [comment]
@benjamingr A PR is added, but I have difficulty explaining the `Promise` returned by the function returned.

**Comment by benjamingr** [comment]
@liushigit thanks! I've added some suggestions, LMK what you think

**Comment by perrin4869** [comment]
I'm not sure if this is the right place to ask this question, but what's the difference performance-wise or other, between using coroutines and async/await in node >= 8? Maybe a note should be added to the documentation about it?

**Comment by benjamingr** [comment]
@perrin4869 it's likely async/await is still slower than `Promise.coroutine` at the moment. You can run the bluebird doxbee benchmarks yourself.
